services/payments/controllers.go

```diff
@@ -0,0 +1,186 @@
+		nonce := make([]byte, 64)
+		_, err := rand.Read(nonce)
+		if err != nil {
+			logger.Error().Err(err).Msg("failed to create random nonce")
+			return handlers.WrapError(err, "failed to create random nonce", http.StatusBadRequest)
+		}
+
+		attestationDocument, err := nitro.Attest(nonce, []byte{}, []byte{})
+		if err != nil {
+			logger.Error().Err(err).Msg("failed to get attestation from nitro")
+			return handlers.WrapError(err, "failed to get attestation from nitro", http.StatusBadRequest)
+		}
+
+		resp := &getConfResponse{
+			// return the attestation document
+			AttestationDocument: base64.StdEncoding.EncodeToString(attestationDocument),
+		}
+
+		logger.Debug().Msg("handling configuration request")
+		// return ok, at this point all new requests will use the new baseCtx of the service
+		return handlers.RenderContent(r.Context(), resp, w, http.StatusOK)
+	})
+}
+
+// PrepareHandler - handler to get current relative exchange rates
+func PrepareHandler(service *Service) handlers.AppHandler {
+	return handlers.AppHandler(func(w http.ResponseWriter, r *http.Request) *handlers.AppError {
+		// get context from request
+		ctx := r.Context()
+
+		var (
+			logger = logging.Logger(ctx, "PrepareHandler")
+			req    = new(Transaction)
+		)
+
+		// read the transactions in the body
+		err := requestutils.ReadJSON(ctx, r.Body, &req)
+		if err != nil {
+			return handlers.WrapError(err, "Error in request body", http.StatusBadRequest)
+		}
+
+		// if dryrun and prepare we will fail here
+		if req.DryRun != nil {
+			if *req.DryRun != prepareFailure {
+				// return a success
+				return handlers.RenderContent(r.Context(), req, w, http.StatusOK)
+			}
+			// return a failure
+			return handlers.WrapError(errors.New("dry run forced error"), "Error in request body", http.StatusBadRequest)
+		}
+
+		logger.Debug().Str("request", fmt.Sprintf("%+v", req)).Msg("structure of request")
+		// validate the transaction
+
+		_, err = govalidator.ValidateStruct(req)
+		if err != nil {
+			logger.Error().Err(err).Str("request", fmt.Sprintf("%+v", req)).Msg("failed to validate structure")
+			return handlers.WrapError(err, "failed to validate transaction", http.StatusBadRequest)
+		}
+
+		// sign the transaction
+		req.PublicKey, req.Signature, err = req.SignTransaction(ctx)
+		if err != nil {
+			logger.Error().Err(err).Str("request", fmt.Sprintf("%+v", req)).Msg("failed to sign transaction")
+			return handlers.WrapError(err, "failed to sign transaction", http.StatusBadRequest)
+		}
+
+		logger.Debug().Str("transaction", fmt.Sprintf("%+v", req)).Msg("handling prepare request")
+
+		// returns an enriched list of transactions, which includes the document metadata
+		resp, err := service.InsertTransaction(ctx, req)
+		if err != nil {
+			return handlers.WrapError(err, "failed to insert transactions", http.StatusInternalServerError)
+		}
+
+		// create a random nonce for nitro attestation
+		nonce := make([]byte, 64)
+		_, err = rand.Read(nonce)
```

Comment:
I believe this nonce is meant to be generated by the client rather than the server in order to create a challenge-response protocol. Otherwise, a malicious server could replay the attestation response back to the CLI tool. In this case, I think that would still be caught based on the Bat Sum check (it wouldn't match the prepare call), but in any case we should be implementing this correctly.

Reply:
See here for how we're doing it in P3A: https://github.com/brave/brave-core/blob/79427585af2140ab15641a3884b4cd3e86c49fde/components/p3a/nitro_utils/attestation.cc#L243-L267